### PR TITLE
Reset array scope tracking for nested objects

### LINKED ISSUE #114873: [CI] LogsDbVersusLogsDbReindexedIntoStandardModeChallengeRestIT testTermsQuery failing
**Build Scans:**
- [elasticsearch-periodic #4433 / openjdk21_checkpart1_java-matrix](https://gradle-enterprise.elastic.co/s/2sjd6ujy5lyqs)
- [elasticsearch-intake #10996 / part1](https://gradle-enterprise.elastic.co/s/4m27fwuhdno6c)

**Reproduction Line:**
```
./gradlew ":modules:data-streams:javaRestTest" --tests "org.elasticsearch.datastreams.logsdb.qa.LogsDbVersusLogsDbReindexedIntoStandardModeChallengeRestIT.testTermsQuery" -Dtests.seed=241394AB18AB51F -Dtests.locale=xog-Latn-UG -Dtests.timezone=Iceland -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testTermsQuery'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.datastreams.logsdb.qa.LogsDbVersusLogsDbReindexedIntoStandardModeChallengeRestIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
org.elasticsearch.client.ResponseException: method [POST], host [http://[::1]:32823], URI [/_reindex?refresh=true], status line [HTTP/1.1 400 Bad Request]
{"took":2766,"timed_out":false,"total":110,"updated":0,"created":109,"deleted":0,"batches":1,"version_conflicts":0,"noops":0,"retries":{"bulk":0,"search":0},"throttled_millis":0,"requests_per_second":-1.0,"throttled_until_millis":0,"failures":[{"index":".ds-standard-apache-reindexed-contender-2024.10.16-000001","id":"98","cause":{"type":"illegal_argument_exception","reason":"mapper [eVob.zkgBrCXeM.shpNltp.RqhGfiWa] cannot be changed from type [dense_vector] to [float]"},"status":400,"failure_store":"not_enabled"}]}
```

**Issue Reasons:**
- [main] 2 failures in test testTermsQuery (0.4% fail rate in 496 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

### PR DESCRIPTION
**Pull request by kkrik-es**
Nested objects are not part of source tracking, as their sole purpose is to properly track object contents within arrays. Array scope tracking needs to be therefore reset when entering their scope, to avoid using the second doc parsing pass for ignored source retrieval on their contents.

Fixes #114873

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java [hunk 1]
```diff
@@ -709,12 +709,14 @@ public final DocumentParserContext createNestedContext(NestedObjectMapper nested
...
      */
     public final DocumentParserContext switchDoc(final LuceneDocument document) {
-        return new Wrapper(this.parent, this) {
+        DocumentParserContext cloned = new Wrapper(this.parent, this) {
             @Override
             public LuceneDocument doc() {
                 return document;
             }
         };
+        cloned.inArrayScope = false;  // Reset for new document.
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java:718]: Maybe add the reason why to reset array scope for nested docs?

**Review by martijnvg — APPROVED** [review]
Good find 👍 

and maybe add a unit test that demonstrates  the problem?

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java (hunk 1 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java]: Updated. 

The second pass doesn't work for nested objects in general, as it adds ignored source to the wrong doc.. It should be extremely rare to need it, it'd be an array within an array within a nested object, and a leaf tracked in ignored source.. 

Updating nested docs in the second pass would be very tricky. We should revisit this if we start tracking parsing scopes.

**Comment by kkrik-es** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/114906)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**